docker/runtime/nodejs/kubeless.js

```diff
@@ -86,7 +88,15 @@ function modExecute(handler, req, res, end) {
             data,
             'extensions': { request: req, response: res },
         };
-        Promise.resolve(func(event, context))
+
+        const funcResult = func(event, context);
```

Comment:
what you are trying to do is better handled in `modFinalize` below. In the case `'object'` you can see if it's a stream (you also don't need the external dependency):

```
function modFinalize(result, res, end) {
    switch(typeof result) {
        case 'string':
            res.end(result);
            break;
        case 'object':
           if (typeof stream.pipe === 'function') {
            result.pipe(res)
           } else {
              res.json(result);
            }
            break;
        default:
            res.end(JSON.stringify(result));
    }
    end();
}

```

Reply:
> what you are trying to do is better handled in modFinalize below. In the case 'object' you can see if it's a stream (you also don't need the external dependency):

Nice, thanks @andresmgot - I'll make some changes and work on a test 👍

Reply:
wouldn't `result instanceof Readable`  be more explicit?

Reply:
I think there is a problem with doing that if the consumer is using https://github.com/nodejs/readable-stream ?

Reply:
Ok, valid point. I was not aware of this package.

Final question:
With the proposed new implementation the stream may end asynchronously, So the runtime actually has to wait for streams `finish` event. Otherwise, the runtime could assume the function is done and report this to Prometheus, even if the stream is actually still piping. See last line in `modFinalize`.

Reply:
Hi @grahamscott  @andresmgot ,

It is a valid requirement that a function uses streams, and of course a `Readable` stream to create the result. However, as grahamscott also pointed out there is not only the core class `Readable` (and derived `Duplex`or `Transform`). There are also other classes from other packages in different versions. And streams may also run in object mode or in chunk mode.

All checks that are needed here will probably be more complex on the core side than on the function side, which knows better what it does. And/or kubeless core would need to have more dependencies which is also not nice.

So here is one further proposal that may fit all:
```
function modFinalize(result, res, end) {
    switch(typeof result) {
        case 'string':
            res.end(result);
            break;
        case 'object':
            if (result === null) {
                res.end();    // if not already ended, TODO
            } else {
                res.json(result);
            }
            break;
        case 'undefined':
            res.end();    // if not already ended, TODO
            break;
        default:
            res.end(JSON.stringify(result));
    }
    end();
}
```
and the function code somehow like this:
```
module.exports = function(event, context) {
    return new Promise((resolve, reject) => {
        ...
        event.extension.res.once('finish', resolve);
        mystream.pipe(event.extension.res, {end: true});
        ...
        mystream.write(data);
        mystream.write(data);
        ...
        mystream.end();
    });
};
```

There is one TODO: find out in `modFinalize` whether or not `res.end()` has already been called, which is needed in any case, independent from the stream use case. Most simple might be a dynamic method redefinition before calling the function code or to observe the `finish` event. Or does express provide the information somehow already?

The copde snippets shall just describe the idea, it is not tested code.

What do you think?

Reply:
Hi @cscheiber 

Thank you for spending time on this, it's much appreciated.

I am a bit worried that it'll add a lot of complexity into core if we don't use some helper libs. There's already Legacy Streams, Streams2 and Streams3 in node, so it gets a bit awkward writing code that targets all three. Hence my selection of the `end-of-stream` dependency. This is only going to introduce 100-120 lines of new code as it'll introduce a dependency tree of only 3 small modules. But, if staying dependency-light in core is a big concern, I am happy to try to figure something else out.

With regards to your example, I'm pretty sure we don't need to call `res.end()` given that we're using `pipe`. We'd also need to think about what happens with problems that cause the stream to break, which is another feather in `end-of-stream`s cap, as it takes care of this for us.

I think it might be nice to push as much of the above into core as possible, so that users creating functions that use streams don't have to worry about as much, but again am happy to go with your decision.

Let me know what you think! I'd love to get this functionality into core one way or another.

Cheers

Reply:
after seeing the alternatives, my take is to make it simple for the core. That way:

 - We don't assume people is going to use streams in a certain way.
 - We avoid third party libraries that people may not need or don't want to add.
 - (less important) We reduce differences with other languages that won't have support for streams.

Let's add documentation for the moment on how to handle streams and make sure we can support both simple or more complex scenarios just with code in the function. If we keep having this request we can implement this in the core or, as the last option, have a different variant of the NodeJS runtime with that feature.

Thank you both for taking the time to review this!

Reply:
Hi @andresmgot 

The way forward isn't so clear for me - we'll need some changes in core in order for streams to work, as otherwise `modFinalize` will cause problems in its current state.

Let me know the kind of scope of changes that you'll consider for core and I'll have a shot at implementing them!

Reply:
Yes, I mean adding the minimum changes in modFinalize needed. As far as I understood the only piece needed is this part:

```
            if (result === null) {
                res.end();    // if not already ended, TODO
            } else {
```

The `Promise` that the function returns should only call `resolve` when the `pipe` that writes in the response object (`event.extension.res`) has finished. The only thing that is not clear is that `modFinalize` should call `res.end()` or not in that case. We should check that.

Reply:
> The only thing that is not clear is if modFinalize should call res.end() or not in that case. We should check that.

I think that we do not need to call `res.end()` explicitly:

From looking at [how Express creates the response object](https://github.com/expressjs/express/blob/master/lib/response.js#L42), it is an instance of [http.ServerResponse](https://nodejs.org/dist/latest-v10.x/docs/api/http.html#http_class_http_serverresponse) and as such implements the Writable Stream interface. So, I think that `res.end()` is an impelmentation of `writable.end()`.

From looking at the [`readable.pipe()` docs](https://nodejs.org/dist/latest-v10.x/docs/api/stream.html#stream_readable_pipe_destination_options), it states that 

>By default, stream.end() is called on the destination Writable stream when the source Readable stream emits 'end', so that the destination is no longer writable. 

So, I think we're good to go!

Reply:
Hi @grahamscott,

Streams are important this is not questioned, latest when it comes to transfer of huge files or the like. It is good that you push this point.

However, with the minimal solution we can enable functions to use streams (Legacy Streams, Streams2 and Streams3 or even the core stream classes) without any assumptions or dependencies in the core, exactly without the things you also want to get rid of with the other helper libs. With the simple approach the libs are not needed at all.

Remark: the core must assure that the response is closed before the function execution is reported as completed. And core cannot rely on a proper function implementation. This usage of `pipe` does not guarantee that data are written or that `end` is really called.

The sample function code is simple, isn't it. You may also add `mystream.on('error', reject)`. And soon we reach applications area for a more fine-grained error handling.

Regarding `res.end()`: the original function could be replaced by an own local function that calls the original function/method but does the job only one time. This helps also in non-stream scenarios.

Reply:
Ok I think I see what you're getting at.
So, how about this:

```
function modFinalize(result, res, end) {
    switch(typeof result) {
        case 'string':
            res.end(result);
            break;
        case 'object':
            if(!res.finished){
                res.json(result);
            }
            break;
        default:
            res.end(JSON.stringify(result));
    }
    end();
}
```

Reply:
`finished` would be nice, but I could find it, even in the latest nodejs core documentation (actually we need to support LTS 6 + 8): https://nodejs.org/docs/latest-v9.x/api/stream.html

And even if we would admit to use `express` there is no such property as well, but please double-check: http://expressjs.com/de/4x/api.html#res

We need also to handle `undefined`. Probably, the whole switch statement would not be done if res is already finished. Maybe, our own `end()` method could set an `finished` property.

Reply:
@cscheiber `finished` is a property of Node's `http.serverResponse`, and has existed since v0.0.2: https://nodejs.org/dist/latest-v10.x/docs/api/http.html#http_response_finished

Express constructs a response object from the prototype of `http.serverResponse`: https://github.com/expressjs/express/blob/master/lib/response.js#L42

So, I think `finished` is fine to use?

Reply:
Sorry, I should have checked also `http.ServerResponse`, you are right, `finished` is there also in v6 and express inherits:
https://nodejs.org/docs/latest-v6.x/api/http.html#http_response_finished

Then there is a nice solution.